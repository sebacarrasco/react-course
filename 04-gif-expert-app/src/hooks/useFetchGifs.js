import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
    // Custom hook que se encarga de hacer la request de las imÃ¡genes
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // Importante hacer notar que los efectos no pueden ser async
    useEffect( () => {
        getGifs(category).then(imgs => {
            setState({
                data: imgs,
                loading: false
            });
        });
    }, [category])

    return state;
}