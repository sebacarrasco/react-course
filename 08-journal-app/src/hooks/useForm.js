import { useState } from "react";

export const useForm = (initialState = {}) => {
    

    const [values, setValues] = useState(initialState);

    const reset = (newFormState = initialState) => {
        setValues(newFormState);
        console.log("hola")
    }

    const handleInputChange = ( { target } ) => {
        setValues({
            ...values,
            [target.name]: target.value
        });
    };

    return [ values, handleInputChange, reset];

};
