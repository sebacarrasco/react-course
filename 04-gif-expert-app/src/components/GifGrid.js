import React from 'react'
import {GifGridItem} from './GifGridItem';
// import { getGifs } from '../helpers/getGifs';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

export const GifGrid = ( { category } ) => {

    // const [images, setImages] = useState([]);

    // // // Cuando cambie un elemento del arreglo que se le pasa
    // // // al useEffect, entonces se volver치 a aejecutar getGifs
    // // // (En este caso solo se ejecutar치 solo una vez ya que
    // // // category no cambia nunca)
    // useEffect( () => {
    //     getGifs(category)
    //         .then(setImages);
    // }, [category]);

    // Lo que est치 comentado arriba es equivalente a lo siguiente
    // (usando un custom hook que por detr치s usa useEffect)
    const { data:images, loading } = useFetchGifs(category);

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>
            {loading && <p>Loading...</p>}
            <div className="card-grid">
                    {
                    images.map(img => 
                            <GifGridItem key={img.id} {...img}/> // Estamos haciendo spread del objeto img para
                                                                // para pasar sus propiedades "sueltas"
                    )
                    }
            </div>
        </>
    )
};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
};