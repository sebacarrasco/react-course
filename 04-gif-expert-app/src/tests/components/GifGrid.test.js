import React from 'react';
import { shallow } from "enzyme";
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';

// Vamos a simular que el custom hook useFetchGifs retornó
// Cone esto se simulará cualquier llamada a ese archivo y
// suponer o controlar la info que esto va a responder
jest.mock('../../hooks/useFetchGifs');

describe('Tests for <GifGrid />', () => {

    const category = "South Park";
    
    test('should match to initial snapshot', () => {

        // Vamos a fingir que el custom hook retorna esto
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        
        const wrapper = shallow(<GifGrid category={ category }/>);

        expect(wrapper).toMatchSnapshot();

    });

    test('should show items when the images are loaded', () => {

        const gifs = [
            {
                id: 1,
                url: "https://example.com",
                title: "A title"
            }
        ];
        
        // Vamos a fingir que el custom hook retorna esto
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        
        const wrapper = shallow(<GifGrid category={ category }/>);

        // expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("p").exists()).toBe(false); // No debería existir el párrafo de loading
        expect(wrapper.find("GifGridItem").length).toBe(gifs.length);


    });
    
    
});