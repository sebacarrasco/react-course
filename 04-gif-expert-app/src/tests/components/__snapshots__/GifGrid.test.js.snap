// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <GifGrid /> should match to initial snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    South Park
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
